### PR TITLE
Infinite ajax loop with undefined IDs

### PR DESCRIPTION
**Pull request by sylwit**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | In some cases IDsArray could return ['undefined'] which produce ajax calls to `/?id_products%5B%5D=undefined`<br>This URI is invalid so it returns the full HTML page and enter in the infinite loop..
| Type?         | bug fix
| BC breaks?    |  no
| Deprecations? | no

### CODE REVIEW

File: views/js/productListingComments.js [hunk 1]
```diff
@@ -87,7 +87,7 @@ var productListingComments = (function () {
         });
 
 
-        var IDsArray = Object.keys(seenIds);
+        var IDsArray = Object.keys(seenIds).filter(e => e !== 'undefined');
```
Comment (NeOMakinG) [inline views/js/productListingComments.js:90]:
```suggestion
        var IDsArray = Object.keys(seenIds).filter(e => typeof e !== 'undefined');
```

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

File: views/js/productListingComments.js (hunk 1 above)
Comment (sylwit) [reply views/js/productListingComments.js:90]: when I checked e was a string, so typeof e was returning string

Comment (NeOMakinG) [reply views/js/productListingComments.js:90]: Well, then this come from the backend

**Review by NeOMakinG — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by lmeyer1** [comment]
What are you waiting on ? This should really be fixed.

**Comment by PierreRambaud** [comment]
> What are you waiting on ? This should really be fixed.

We are trying to find a way to reproduce the issue, unfortunately, we have not enough information :( If you can help us in order to test and validate it? :)

**Review by lmeyer1 — COMMENTED** [review]
The problem comes from line 85 and 86
```
var id = $(this).data('id');
seenIds[id] = true;
```
If `typeof id === 'undefined'`, when used as an index, it is cast to the string `undefined`

A better correction would be on line 86
`if (typeof id !== 'undefined') seenIds[id] = true;`
instead of the proposed check a posteriori on line 90.

**Comment by PierreRambaud** [comment]
> The problem comes from line 85 and 86
> 
> ```
> var id = $(this).data('id');
> seenIds[id] = true;
> ```
> 
> If `typeof id === 'undefined'`, when used as an index, it is cast to the string `undefined`
> 
> A better correction would be on line 86
> `if (typeof id !== 'undefined') seenIds[id] = true;`
> instead of the proposed check a posteriori on line 90.

I agree, but the problem is, how `$(this).data('id');` can returns undefined :sweat_smile: 
What's wrong with the selector :thinking:

**Comment by lmeyer1** [comment]
There's nothing wrong with the selector itself. This line just requires a them to include `data-id="[product-id]"` in the product item template.
Now, if the template does not follow this requirement, we get an infinite loop.
When approving this PR, when the template does not satisfy the requirement, instead of an infinite loop, it fails silently, what is a much better behavior.

**Comment by PierreRambaud** [comment]
I'm merging it as the QA team doesn't have time to check it, and you only add a verification :+1: 
Thank you @sylwit

**Pull request merged**